### ISSUE TITLE
Partial access to block shared memory

### ISSUE THREAD

**Issue by jklmnn** [Componolit/gneiss#115]
The current block interface accessors only allow fully reading or writing the shared memory region assigned to a block request. This limits the maximum size of block requests that can be handled (especially by the server) to the amount of internally available memory.
Furthermore in the client the platform decides when the application can write the required memory into write requests.

# Granularity

The callback interfaces of client and server only pass the shared memory by reference so the application is free to access any region it desires (or to not access a region).
The direct interface of the block server also allows to read/write the required data directly via an `in out` parameter. Yet this parameter is required to be complete so each time the whole request has to be written/read. To solve this the interface should allow byte wise access.
The new interface will look as follows:
```Ada
procedure Read (S : in out Server_Session;
                R :        Request;
                B :        Buffer;
                F :        Buffer_Index;
                L :        Buffer_Index);

procedure Write (S : in out Server_Session;
                 R :        Request;
                 B :    out Buffer;
                 F :        Buffer_Index;
                 L :        Buffer_Index);
```
Both procedures will receive an additional first (`F`) and last (`L`) parameter that designate the range that is to be accessed. This range is byte granular within the buffer allowing to fill even single blocks partially. The application decides when it finished accessing the buffer by acknowledging the request.

# Decision

In both server and client the application can decide when to write or read request memory, and how often to do so. And it has a possibility to decide when it is ready to stop accessing the memory (Acknowledge/Release).
The only exception is the call to `Write` in the client which is currently triggered when the client calls `Enqueue`. As the request is enqueued after this call and cannot be accessed anymore the client is in pressure to provide all required data before returning from `Write`. This has two problems. The client cannot craft requests larger than the continuous memory it can hold by itself. Also if the enqueuing failed on the platform the `Write` procedure will eventually be called again.
To solve this the write will be triggered similarly to the read by calling a distinct procedure on the request that will then call the callback:
```Ada
procedure Write (C : in out Client_Session;
                 R :        Request);
```
The procedure will be only callable in the `Allocated` state (the request has been allocated, so the memory is available but it has not yet been given to the platform with `Enqueue`.

**Comment by jklmnn** [comment]
Since the length of the buffer is already given via the buffer itself it is enough to provide an offset from the start of the assigned shared memory region (example for `Read` only as `Write` is analogous):
```Ada
procedure Read (S : in out Server_Session;
                R :        Request;
                B :        Buffer;
                O :        Byte_Length);
```

**Comment by jklmnn** [comment]
Fixed by #116.
